pkg/collector/corechecks/snmp/discovery/discovery.go

```diff
@@ -0,0 +1,294 @@
+package discovery
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/persistentcache"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/snmp/checkconfig"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/snmp/devicecheck"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/snmp/session"
+)
+
+const cacheKeyPrefix = "snmp_corecheck"
+const sysObjectIDOid = "1.3.6.1.2.1.1.2.0"
+
+// Discovery handles snmp discovery states
+type Discovery struct {
+	config            *checkconfig.CheckConfig
+	stop              chan struct{}
+	discDevMu         sync.RWMutex
+	discoveredDevices map[string]Device
+}
+
+// Device implements and store results from the Service interface for the SNMP listener
+type Device struct {
+	entityID    string
+	deviceIP    string
+	deviceCheck *devicecheck.DeviceCheck
+}
+type snmpSubnet struct {
+	config     *checkconfig.CheckConfig
+	startingIP net.IP
+	network    net.IPNet
+
+	cacheKey string
+	devices  map[string]string
+
+	deviceFailures map[string]int
+}
+
+type snmpJob struct {
+	subnet    *snmpSubnet
+	currentIP net.IP
+}
+
+// Start discovery
+func (d *Discovery) Start() {
+	log.Debugf("subnet %s: Start discovery", d.config.Network)
+	go d.checkDevices()
+}
+
+// Stop signal discovery to shut down
+func (d *Discovery) Stop() {
+	log.Debugf("subnet %s: Stop discovery", d.config.Network)
+	close(d.stop)
+}
+
+// GetDiscoveredDeviceConfigs returns discovered device configs
+func (d *Discovery) GetDiscoveredDeviceConfigs() []*devicecheck.DeviceCheck {
+	d.discDevMu.RLock()
+	defer d.discDevMu.RUnlock()
+
+	discoveredDevices := make([]*devicecheck.DeviceCheck, 0, len(d.discoveredDevices))
+	for _, device := range d.discoveredDevices {
+		discoveredDevices = append(discoveredDevices, device.deviceCheck)
+	}
+	return discoveredDevices
+}
+
+// Start discovery
+func (d *Discovery) runWorker(w int, jobs <-chan snmpJob) {
+	log.Debugf("subnet %s: Start SNMP worker %d", d.config.Network, w)
+	for {
+		select {
+		case <-d.stop:
+			log.Debugf("subnet %s: Stop SNMP worker %d", d.config.Network, w)
+			return
+		case job := <-jobs:
+			log.Debugf("subnet %s: Handling IP %s", d.config.Network, job.currentIP.String())
+			err := d.checkDevice(job)
+			if err != nil {
+				log.Errorf(err.Error())
+			}
+		}
+	}
+}
+
+func (d *Discovery) checkDevices() {
```

Comment:
Could this be `findDevices` or `lookForDevices` ? `check` makes me think of the snmp check

Reply:
This could also be `discoverNetwork`